services/spar/src/Spar/SCIM.hs

```diff
@@ -0,0 +1,304 @@
+      , SCIM.lastModified = testDate
+      , SCIM.version = SCIM.Strong (Text.pack (show thingHash))
+      , SCIM.location = SCIM.URI $ URI "https://TODO" Nothing "" "" ""
+      }
+
+-- 2018-01-01 00:00
+--
+-- TODO: real dates!
+testDate :: UTCTime
+testDate = UTCTime
+  { utctDay = ModifiedJulianDay 58119
+  , utctDayTime = 0
+  }
+
+emptySCIMName :: SCIM.User.Name
+emptySCIMName =
+  SCIM.User.Name
+    { SCIM.User.formatted = Nothing
+    , SCIM.User.givenName = Just ""
+    , SCIM.User.familyName = Just ""
+    , SCIM.User.middleName = Nothing
+    , SCIM.User.honorificPrefix = Nothing
+    , SCIM.User.honorificSuffix = Nothing
+    }
+
+{- TODO: might be useful later.
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-- | Parse a name from a user profile into an SCIM name (Okta wants given
+-- name and last name, so we break our names up to satisfy Okta).
+--
+-- TODO: use the same algorithm as Wire clients use.
+toSCIMName :: Name -> SCIM.User.Name
+toSCIMName (Name name) =
+  SCIM.User.Name
+    { SCIM.User.formatted = Just name
+    , SCIM.User.givenName = Just first
+    , SCIM.User.familyName = if Text.null rest then Nothing else Just rest
+    , SCIM.User.middleName = Nothing
+    , SCIM.User.honorificPrefix = Nothing
+    , SCIM.User.honorificSuffix = Nothing
+    }
+  where
+    (first, Text.drop 1 -> rest) = Text.breakOn " " name
+
+-- | Convert from the Wire phone type to the SCIM phone type.
+toSCIMPhone :: Phone -> SCIM.User.Phone
+toSCIMPhone (Phone phone) =
+  SCIM.User.Phone
+    { SCIM.User.typ = Nothing
+    , SCIM.User.value = Just phone
+    }
+
+-- | Convert from the Wire email type to the SCIM email type.
+toSCIMEmail :: Email -> SCIM.User.Email
+toSCIMEmail (Email eLocal eDomain) =
+  SCIM.User.Email
+    { SCIM.User.typ = Nothing
+    , SCIM.User.value = SCIM.User.EmailAddress2
+        (unsafeEmailAddress (encodeUtf8 eLocal) (encodeUtf8 eDomain))
+    , SCIM.User.primary = Just True
+    }
+
+-}
+
+-- Note [error handling]
+-- ~~~~~~~~~~~~~~~~~
+--
+-- There are two problems with error handling here:
+--
+-- 1. We want all errors originating from SCIM handlers to be thrown as SCIM
+--    errors, not as Spar errors. Currently errors thrown from things like
+--    'getTeamMembers' will look like Spar errors and won't be wrapped into
+--    the 'SCIMError' type. This might or might not be important, depending
+--    on what is expected by apps that use the SCIM interface.
+--
+-- 2. We want generic error descriptions in response bodies, while still
+--    logging nice error messages internally.
+
+instance SCIM.UserDB (ReaderT TeamId Spar) where
+  -- | List all users, possibly filtered by some predicate.
+  list mbFilter = do
+    tid <- ask
+    members <- lift $ getTeamMembers tid
+    users <- forM members $ \member ->
+      lift (getUser (member ^. Galley.userId)) >>= \case
+        Just user -> pure (toSCIMUser user)
+        Nothing -> SCIM.throwSCIM $
+          SCIM.serverError "Database is inconsistent"
```

Comment:
Put some more details in here?

```haskell
"DB inconsistent: user " <> show userid <> " exists in galley, but not in brig."
```

I guess we need to worry about this leaking to the UI if we do it, but I think even if that happened it would be preferable to us starting at "you got it all wrong"-style errors.